contracts/proxy-lib/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct ProposalWithSigner {
+    pub proposal: Proposal,
+    pub signer_id: Repr<SignerId>,
+}
+
+// An internal request wrapped with the signer_pk and added timestamp to determine num_requests_pk and prevent against malicious key holder gas attacks
+#[derive(Clone, PartialEq)]
+#[near(serializers = [json, borsh])]
+pub struct ConfirmationRequestWithSigner {
+    pub proposal_id: ProposalId,
+    pub signer_id: Repr<SignerId>,
+    pub added_timestamp: u64,
+}
+
+/// Lowest level action that can be performed by the multisig contract.
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub enum ProposalAction {
+    FunctionCall {
+        method_name: String,
+        args: Base64VecU8,
+        deposit: U128,
+        gas: U64,
+    },
+}
+
+// The request the user makes specifying the receiving account and actions they want to execute (1 tx)
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct Proposal {
+    pub receiver_id: AccountId,
+    pub actions: Vec<ProposalAction>,
+}
+
+#[near]
+impl ProxyContract {
+    #[init]
+    pub fn init(context_id: Repr<ContextId>, context_config_account_id: AccountId) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            context_id,
+            context_config_account_id,
+            proposal_nonce: 0,
+            proposals: IterableMap::new(b"r".to_vec()),
+            approvals: IterableMap::new(b"c".to_vec()),
+            num_proposals_pk: IterableMap::new(b"k".to_vec()),
+            num_approvals: 2,
+            active_proposals_limit: 10,
+        }
+    }
+
+    pub fn create_proposal(&mut self, proposal: Signed<ProposalWithSigner>) -> ProposalId {
+        // Verify the signature corresponds to the signer_id
+        let proposal = proposal
+            .parse(|i| *i.signer_id)
+            .expect("failed to parse input");
+
+        let singer_id = &proposal.signer_id;
+
+        let num_proposals = self.num_proposals_pk.get(singer_id).unwrap_or(&0) + 1;
+        assert!(
+            num_proposals <= self.active_proposals_limit,
+            "Account has too many active proposals. Confirm or delete some."
+        );
+        self.num_proposals_pk
+            .insert(singer_id.clone(), num_proposals);
+
+        self.proposals.insert(self.proposal_nonce, proposal.clone());
+        self.approvals.insert(self.proposal_nonce, HashSet::new());
+        self.proposal_nonce += 1;
+        self.proposal_nonce - 1
+    }
+
+    pub fn create_and_approve_proposal(&mut self, proposal: Signed<ProposalWithSigner>) -> Promise {
+        let proposal_id = self.create_proposal(proposal.clone());
+        let proposal = proposal
+            .parse(|i| *i.signer_id)
+            .expect("failed to parse input");
```

Comment:
`create_proposal` will call `.parse` and so are you here.

you can introduce a non-`pub`* method that creates a proposal from a pre-parsed proposal.

*: I didn't say "private" since `near-sdk` has a concept of `#[private]` and that's not what I mean